### PR TITLE
Enhance SSL_client_hello_get0_ext to support all extensions

### PR DESCRIPTION
**Pull request by mshekiwy**
##### Checklist

- [ ] documentation is added or updated
- [ ] tests are added or updated

### CODE REVIEW

File: ssl/ssl_lib.c [hunk 1]
```diff
@@ -6677,18 +6677,24 @@ int SSL_client_hello_get_extension_order(SSL *s, uint16_t *exts, size_t *num_ext
...
+
+    pkt = sc->clienthello->extensions;
+    while (PACKET_remaining(&pkt) > 0) {
+        if (!PACKET_get_net_2(&pkt, &ext_type) ||
+            !PACKET_get_length_prefixed_2(&pkt, &ext_data))
+		return 0;
+
+        if (ext_type == type) {
             if (out != NULL)
                 *out = PACKET_data(&r->data);
```
Comment (mattcaswell) [inline ssl/ssl_lib.c:6699]: This results in a compilation error because you removed the `r` variable

**Review by mattcaswell — CHANGES_REQUESTED** [review]

**Comment by mattcaswell** [comment]
We will need to see a test added for this. Probably by adding a test case to test/sslapitest.c

File: ssl/ssl_lib.c (hunk 1 above)
Comment (mshekiwy) [reply ssl/ssl_lib.c]: fixed

**Comment by vdukhovni** [comment]
I am curious what specifically motives this PR. In what way are `SSL_CTX_add_client_custom_ext` and `SSL_CTX_add_server_custom_ext` insufficient for your needs?

**Comment by mshekiwy** [comment]
hi @vdukhovni im curious too , how SSL_CTX_add_client_custom_ext and SSL_CTX_add_server_custom_ext can replace SSL_client_hello_get0_ext ?

**Comment by vdukhovni** [comment]
> hi @vdukhovni im curious too , how SSL_CTX_add_client_custom_ext and SSL_CTX_add_server_custom_ext can replace SSL_client_hello_get0_ext ?

They are not direct replacements for bulk retrieval of the full list of extensions, but they do support handling of any custom extensions you define callbacks for one at a time.  So I am curious what the use-case is for bulk retrieval?

Sure, if there's good reason to want them all at once, then fine. But perhaps the extant callback interface is sufficient?
And bulk retrieval will expose the extensions that are either built-in, or for which there are callbacks, and leave out "ignored" extensions? Or it is specifically the "ignored" extensions you want to inspect (for diagnostic or other purposes)?

**Comment by mshekiwy** [comment]
@vdukhovni  the function does not do bulk retrieval of the full list of extensions, it returns the one with type matches .
and i think it is a good point and will save a lot of work not to add a callback or register the extension after adding a new one, 
as long as there is no security issue from doing this,

**Comment by vdukhovni** [comment]
> @vdukhovni the function does not do bulk retrieval of the full list of extensions, it returns the one with type matches . and i think it is a good point and will save a lot of work not to add a callback or register the extension after adding a new one, as long as there is no security issue from doing this,

If this is supposed to be an alternative to registering a callback, it is a not a good choice.  With the callback, OpenSSL is aware that the extension is supported and handled, and you can specify an alert when the extension payload inappropriate.

This API is instead informational, and the SSL engine is not aware of the disposition of the extension. If you're implementing support for a new custom extension, you should use the callback interface.

**Comment by vdukhovni** [comment]
Closing this issue. If there's a more compelling rationale for the new API, please reopen and elaborate.

**Pull request closed without merging**